pkg/agent/endpoints/sdsv3/handler.go

```diff
@@ -307,6 +337,171 @@ func (h *Handler) triggerReceivedHook() {
 	}
 }
 
+type validationContextBuilder interface {
+	buildOne(resourceName, trustDomainID string) (*any.Any, error)
+	buildAll(resourceName string) (*any.Any, error)
+}
+
+type rootCABuilder struct {
+	bundles map[string]*bundleutil.Bundle
+}
+
+func newRootCABuilder(bundle *bundleutil.Bundle, federatedBundles map[spiffeid.TrustDomain]*bundleutil.Bundle) validationContextBuilder {
+	bundles := make(map[string]*bundleutil.Bundle, len(federatedBundles)+1)
+	// Add td bundle if provided
+	if bundle != nil {
+		bundles[bundle.TrustDomainID()] = bundle
+	}
+
+	for td, federatedBundle := range federatedBundles {
+		bundles[td.IDString()] = federatedBundle
+	}
+
+	return &rootCABuilder{
+		bundles: bundles,
+	}
+}
+
+func (b *rootCABuilder) buildOne(resourceName, trustDomain string) (*any.Any, error) {
+	bundle, ok := b.bundles[trustDomain]
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "no bundle found for trust domain: %q", trustDomain)
+	}
+	caBytes := pemutil.EncodeCertificates(bundle.RootCAs())
+	return anypb.New(&tls_v3.Secret{
+		Name: resourceName,
+		Type: &tls_v3.Secret_ValidationContext{
+			ValidationContext: &tls_v3.CertificateValidationContext{
+				TrustedCa: &core_v3.DataSource{
+					Specifier: &core_v3.DataSource_InlineBytes{
+						InlineBytes: caBytes,
+					},
+				},
+			},
+		},
+	})
+}
+
+func (b *rootCABuilder) buildAll(resourceName string) (*any.Any, error) {
+	return nil, status.Error(codes.Internal, `unable to use "SPIFFE validator" on Envoy below 1.17`)
+}
+
+type spiffeBuilder struct {
+	bundles map[spiffeid.TrustDomain]*bundleutil.Bundle
+}
+
+func newSpiffeBuilder(tdBundle *bundleutil.Bundle, federatedBundles map[spiffeid.TrustDomain]*bundleutil.Bundle) (validationContextBuilder, error) {
+	bundles := make(map[spiffeid.TrustDomain]*bundleutil.Bundle, len(federatedBundles)+1)
+
+	// Add td bundle if provided
```

Comment:
FWIW, tdBundle should never be nil. I'm ok with the defensive code, but we should adjust the comment so folks don't think it can be nil in normal circumstances.
```suggestion
	// Only include tdBundle if it is not nil, which shouldn't ever be the case. This is purely defensive.
```